File: backend/core/encryption_service.py
"""
Message Encryption Service for end-to-end encrypted chat
Uses RSA + AES hybrid encryption
"""
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.backends import default_backend
import base64
import os


class MessageEncryptionService:
    """Handles end-to-end encryption for chat messages"""
    
    @staticmethod
    def generate_user_keys():
        """
        Generate RSA key pair for a user
        Returns: (public_key_pem, private_key_pem)
        """
        # Generate 2048-bit RSA key pair
        private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=2048,
            backend=default_backend()
        )
        public_key = private_key.public_key()
        
        # Serialize to PEM format
        private_pem = private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption()  # No password protection here
        ).decode('utf-8')
        
        public_pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        ).decode('utf-8')
        
        return public_pem, private_pem
    
    @staticmethod
    def encrypt_message(content, recipient_public_keys):
        """
        Encrypt message with AES, then encrypt AES key with each recipient's RSA public key
        
        Args:
            content: str - The message content to encrypt
            recipient_public_keys: dict - {user_id: public_key_pem}
        
        Returns:
            dict: {'encrypted_content': base64_str, 'encrypted_keys': {user_id: encrypted_aes_key}}
        """
        # Generate random AES key (256-bit) and IV (128-bit)
        aes_key = os.urandom(32)
        iv = os.urandom(16)
        
        # Pad content to AES block size (16 bytes)
        content_bytes = content.encode('utf-8')
        block_size = 16
        padding_length = block_size - (len(content_bytes) % block_size)
        padded_content = content_bytes + bytes([padding_length] * padding_length)
        
        # Encrypt content with AES-CBC
        cipher = Cipher(
            algorithms.AES(aes_key),
            modes.CBC(iv),
            backend=default_backend()
        )
        encryptor = cipher.encryptor()
        encrypted_content = encryptor.update(padded_content) + encryptor.finalize()
        
        # Combine IV and encrypted content
        encrypted_data = iv + encrypted_content
        encrypted_content_b64 = base64.b64encode(encrypted_data).decode('utf-8')
        
        # Encrypt AES key for each recipient
        encrypted_keys = {}
        for user_id, public_key_pem in recipient_public_keys.items():
            # Load public key
            public_key = serialization.load_pem_public_key(
                public_key_pem.encode('utf-8'),
                backend=default_backend()
            )
            
            # Encrypt AES key with RSA-OAEP
            encrypted_aes_key = public_key.encrypt(
                aes_key,
                padding.OAEP(
                    mgf=padding.MGF1(algorithm=hashes.SHA256()),
                    algorithm=hashes.SHA256(),
                    label=None
                )
            )
            
            encrypted_keys[str(user_id)] = base64.b64encode(encrypted_aes_key).decode('utf-8')
        
        return {
            'encrypted_content': encrypted_content_b64,
            'encrypted_keys': encrypted_keys
        }
    
    @staticmethod
    def decrypt_message(encrypted_content_b64, encrypted_aes_key_b64, private_key_pem):
        """
        Decrypt message using user's private key
        
        Args:
            encrypted_content_b64: str - Base64 encoded encrypted content
            encrypted_aes_key_b64: str - Base64 encoded encrypted AES key
            private_key_pem: str - User's private key in PEM format
        
        Returns:
            str: Decrypted message content
        """
        # Load private key
        private_key = serialization.load_pem_private_key(
            private_key_pem.encode('utf-8'),
            password=None,
            backend=default_backend()
        )
        
        # Decrypt AES key with RSA
        encrypted_aes_key = base64.b64decode(encrypted_aes_key_b64)
        aes_key = private_key.decrypt(
            encrypted_aes_key,
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None
            )
        )
        
        # Decode encrypted content
        encrypted_data = base64.b64decode(encrypted_content_b64)
        iv = encrypted_data[:16]
        encrypted_content = encrypted_data[16:]
        
        # Decrypt content with AES-CBC
        cipher = Cipher(
            algorithms.AES(aes_key),
            modes.CBC(iv),
            backend=default_backend()
        )
        decryptor = cipher.decryptor()
        padded_content = decryptor.update(encrypted_content) + decryptor.finalize()
        
        # Remove PKCS7 padding
        padding_length = padded_content[-1]
        content = padded_content[:-padding_length]
        
        return content.decode('utf-8')

File: consultancy-dev/lib/encryptionService.ts
/**
 * Client-Side Encryption Service for End-to-End Encrypted Chat
 * Uses RSA + AES hybrid encryption matching the backend implementation
 * 
 * NOTE: For production use, RSA key generation should happen server-side
 * with proper password-based encryption of private keys. This is a simplified
 * implementation for demonstration.
 */

import CryptoJS from 'crypto-js';
import forge from 'node-forge';

export class ClientEncryptionService {
    /**
     * Generate RSA key pair for a user (2048-bit)
     * In production, this should be done server-side with password protection
     */
    static generateKeyPair(): { publicKey: string; privateKey: string } {
        const keypair = forge.pki.rsa.generateKeyPair({ bits: 2048, e: 0x10001 });

        return {
            publicKey: forge.pki.publicKeyToPem(keypair.publicKey),
            privateKey: forge.pki.privateKeyToPem(keypair.privateKey)
        };
    }

    /**
     * Encrypt a message with AES, then encrypt the AES key with recipients' RSA public keys
     * This matches the backend MessageEncryptionService.encrypt_message
     * 
     * @param content - Plain text message  
     * @param recipientPublicKeys - Map of userId to PEM public key
     * @returns Object with encrypted_content and encrypted_keys
     */
    static encryptMessage(
        content: string,
        recipientPublicKeys: Record<string, string>
    ): { encrypted_content: string; encrypted_keys: Record<string, string> } {
        // Generate random AES key (256-bit) and IV (128-bit)
        const aesKey = CryptoJS.lib.WordArray.random(32); // 32 bytes = 256 bits
        const iv = CryptoJS.lib.WordArray.random(16); // 16 bytes = 128 bits

        // Encrypt content with AES-CBC
        const encrypted = CryptoJS.AES.encrypt(content, aesKey, {
            iv: iv,
            mode: CryptoJS.mode.CBC,
            padding: CryptoJS.pad.Pkcs7
        });

        // Combine IV and encrypted content (matching backend format)
        const encryptedData = iv.concat(encrypted.ciphertext);
        const encryptedContentB64 = CryptoJS.enc.Base64.stringify(encryptedData);

        // Encrypt AES key for each recipient with their RSA public key
        const encryptedKeys: Record<string, string> = {};

        Object.entries(recipientPublicKeys).forEach(([userId, publicKeyPem]) => {
            try {
                // Load public key from PEM
                const publicKey = forge.pki.publicKeyFromPem(publicKeyPem);

                // Convert AES key to bytes
                const aesKeyBytes = forge.util.createBuffer(
                    CryptoJS.enc.Hex.parse(aesKey.toString()).toString(CryptoJS.enc.Latin1)
                ).getBytes();

                // Encrypt AES key with RSA-OAEP (matching backend)
                const encryptedAesKey = publicKey.encrypt(aesKeyBytes, 'RSA-OAEP', {
                    md: forge.md.sha256.create(),
                    mgf1: {
                        md: forge.md.sha256.create()
                    }
                });

                // Base64 encode the encrypted AES key
                encryptedKeys[userId] = forge.util.encode64(encryptedAesKey);
            } catch (error) {
                console.error(`Failed to encrypt for user ${userId}:`, error);
                throw new Error(`Failed to encrypt message for recipient ${userId}`);
            }
        });

        return {
            encrypted_content: encryptedContentB64,
            encrypted_keys: encryptedKeys
        };
    }

    /**
     * Decrypt a message using user's private key
     * This matches the backend MessageEncryptionService.decrypt_message
     * 
     * @param encryptedContentB64 - Base64 encoded encrypted content (IV + ciphertext)
     * @param encryptedAesKeyB64 - Base64 encoded encrypted AES key for this user
     * @param privateKeyPem - User's private key in PEM format
     * @returns Decrypted plain text message
     */
    static decryptMessage(
        encryptedContentB64: string,
        encryptedAesKeyB64: string,
        privateKeyPem: string
    ): string {
        try {
            // Load private key from PEM
            const privateKey = forge.pki.privateKeyFromPem(privateKeyPem);

            // Decrypt AES key with RSA
            const encryptedAesKey = forge.util.decode64(encryptedAesKeyB64);
            const aesKeyBytes = privateKey.decrypt(encryptedAesKey, 'RSA-OAEP', {
                md: forge.md.sha256.create(),
                mgf1: {
                    md: forge.md.sha256.create()
                }
            });

            // Convert decrypted AES key bytes to CryptoJS WordArray
            const aesKeyHex = forge.util.bytesToHex(aesKeyBytes);
            const aesKey = CryptoJS.enc.Hex.parse(aesKeyHex);

            // Decode encrypted content (IV + ciphertext)
            const encryptedData = CryptoJS.enc.Base64.parse(encryptedContentB64);

            // Extract IV (first 16 bytes) and ciphertext (rest)
            const iv = CryptoJS.lib.WordArray.create(encryptedData.words.slice(0, 4)); // 4 words = 16 bytes
            const ciphertext = CryptoJS.lib.WordArray.create(encryptedData.words.slice(4));

            // Decrypt content with AES-CBC
            const decrypted = CryptoJS.AES.decrypt(
                { ciphertext } as any,
                aesKey,
                {
                    iv: iv,
                    mode: CryptoJS.mode.CBC,
                    padding: CryptoJS.pad.Pkcs7
                }
            );

            // Convert to UTF-8 string
            return decrypted.toString(CryptoJS.enc.Utf8);
        } catch (error) {
            console.error('Decryption failed:', error);
            throw new Error('Failed to decrypt message');
        }
    }

    /**
     * Helper: Check if user has encryption keys set up
     */
    static hasKeys(publicKey?: string | null, privateKey?: string | null): boolean {
        return Boolean(publicKey && privateKey);
    }

    /**
     * Helper: Store keys in localStorage (NOT SECURE - for demo only)
     * In production, private keys should NEVER be stored in localStorage
     * They should be encrypted with user's password and stored server-side
     */
    static storeKeys(publicKey: string, privateKey: string) {
        if (typeof window !== 'undefined') {
            localStorage.setItem('chat_public_key', publicKey);
            // WARNING: Storing private key in localStorage is NOT secure
            // This is for demonstration only
            localStorage.setItem('chat_private_key', privateKey);
        }
    }

    /**
     * Helper: Retrieve keys from localStorage
     */
    static getStoredKeys(): { publicKey: string | null; privateKey: string | null } {
        if (typeof window !== 'undefined') {
            return {
                publicKey: localStorage.getItem('chat_public_key'),
                privateKey: localStorage.getItem('chat_private_key')
            };
        }
        return { publicKey: null, privateKey: null };
    }
}
